cunumeric/deferred.py

```diff
@@ -305,37 +305,288 @@ def get_scalar_array(self):
+            for dim, k in enumerate(key):
+                if np.isscalar(k) or isinstance(k, NumPyThunk):
+                    if start_index == -1:
+                        start_index = dim
+                    transpose_indices += (dim,)
+                    transpose_needed = transpose_needed or (
+                        (dim - last_index) > 1
+                    )
+                    last_index = dim
+
+            if transpose_needed:
+                copy_needed = True
+                start_index = 0
+                post_indices = tuple(
+                    i for i in range(store.ndim) if i not in transpose_indices
+                )
+                transpose_indices += post_indices
+                store = store.transpose(transpose_indices)
+                index_map = list(transpose_indices)
+                count = 0
+                for i in transpose_indices:
+                    index_map[i] = count
+                    count += 1
+            else:
+                index_map = tuple(range(len(key)))
+
+            for d, k in enumerate(key):
+                dim = index_map[d]
+                if np.isscalar(k):
+                    if k < 0:
+                        k += store.shape[dim + shift]
+                    store = store.project(dim + shift, k)
+                    shift -= 1
+                    copy_needed = True
+                elif k is np.newaxis:
+                    store = store.promote(dim + shift, 1)
+                    copy_needed = True
+                elif isinstance(k, slice):
+                    store = store.slice(dim + shift, k)
+                    if k != slice(None):
+                        copy_needed = True
+                elif isinstance(k, NumPyThunk):
+                    if k.dtype == np.bool:
+                        if k.shape[0] != store.shape[dim]:
+                            raise ValueError(
+                                "boolean index did not match "
+                                "indexed array along dimension  "
+                            )
+                        # in case of the mixed indises we all nonzero
+                        # for the bool array
+                        k = k.nonzero()
+                        tuple_of_arrays += k
+                    else:
+                        tuple_of_arrays += (self.runtime.to_deferred_array(k),)
+                else:
+                    raise TypeError(
+                        "Unsupported entry type passed to advanced",
+                        "indexing operation",
                     )
-                tuple_of_arrays += (k,)
+            if copy_needed or (not store._transform.bottom):
+                # after store is transformed we need to to return a copy of
+                # the store since Copy operation can't be done on
+                # the store with transformation
+                rhs, store = self.copy_store(store)
         else:
             assert isinstance(key, NumPyThunk)
```

Comment:
You should convert to a `DeferredArray` at this point. Otherwise you can't use `key.base`, as you do later in this branch.

Reply:
You can't *necessarily* use `key.base`, because not all subclasses of `NumPyThunk` have a `Store`-type `.base`, only `DeferredArray` (e.g. `EagerArray` doesn't have a `.base`).

Reply:
Makes sense , I will replace it with `assert isinstance(key, DeferredArray)`

Reply:
Actually this is not sufficient. The index array could also be an `EagerArray`. For example, if you run without `-cunumeric:test`, and the base array is large enough to become a `DeferredArray`, while the index array is small enough that it becomes an `EagerArray`. This code:

```
x = cn.ones((300,400))
print(x[[1,1]])
```

fails with:

```
  File "/Users/mpapadakis/legate.core/install/lib/python3.7/site-packages/legion_top.py", line 410, in legion_python_main
    run_path(args[start], run_name='__main__')
  File "/Users/mpapadakis/legate.core/install/lib/python3.7/site-packages/legion_top.py", line 234, in run_path
    exec(code, module.__dict__, module.__dict__)
  File "a.py", line 7, in <module>
    print(x[[1,1]])
  File "/Users/mpapadakis/cunumeric/cunumeric/array.py", line 139, in wrapper
    return func(*args, **kwargs)
  File "/Users/mpapadakis/cunumeric/cunumeric/array.py", line 781, in __getitem__
    return ndarray(shape=None, thunk=self._thunk.get_item(key))
  File "/Users/mpapadakis/cunumeric/cunumeric/deferred.py", line 643, in get_item
    copy_needed, rhs, index_array = self._create_indexing_array(key)
  File "/Users/mpapadakis/cunumeric/cunumeric/deferred.py", line 527, in _create_indexing_array
    assert isinstance(key, DeferredArray)
AssertionError
```

You need to check that it's a `NumPyThunk`, as you were doing before, and if so convert to `DeferredArray`.

Reply:
Will do, thank you